### PR TITLE
Add minimum stakes limit

### LINKED ISSUE #89: StakingPool: add minimum stakes limit
To avoid malicious attacks.

### PR DESCRIPTION
**Pull request by jishankai**
Resolves #89

### CODE REVIEW

**Comment by jishankai** [comment]
> Couple of questions:
> 1, A user can bypass the min stake requirements by staking n and then withdrawing n-x so that it could stake x, which can be smaller than the limit. How could we solve the problem?
> 2, Who is able to change the minimal stake?

1. Add minStakes check in withdrawStakes
2. No one can change minimal stakes after this contract is deployed, it's the same as maxStakers.

File: contracts/StakingPool.sol [hunk 1]
```diff
@@ -99,6 +103,8 @@ contract StakingPool {
         StakerInfo storage info = stakerInfo[msg.sender];
         assert(stakers[info.arrPos] == msg.sender);
         require(info.stakes >= amount, "Should have enough stakes to withdraw.");
+        // solium-disable-next-line max-len
+        require(info.stakes.sub(amount) == 0 || info.stakes.sub(amount) >= minStakes, "Should satisfy minimum stakes.");
```
Comment (qizhou) [inline contracts/StakingPool.sol:107]: I guess we could avoid the warnining by writing the code like:
require(
    info.stakes.sub(amount) == 0 ||
    info.stakes.sub(amount) >= minStakes, "Should satisfy minimum stakes."
);

E.g., https://github.com/QuarkChain/quarkchain-contracts/blob/541b667fd953c942f7608d7e8c71e58b9c3dfff0/contracts/GeneralNativeTokenManager.sol#L134

Comment (jishankai) [reply contracts/StakingPool.sol]: Thanks a lot for the suggestion.

File: contracts/StakingPool.sol [hunk 2]
```diff
@@ -80,6 +83,7 @@ contract StakingPool {
 
     // Add stakes
     function () external payable {
+        require(minStakes <= msg.value, "Invalid stakes.");
```
Comment (freedostudio) [inline contracts/StakingPool.sol:86]: Should require info.stakes >= minStakes

File: contracts/StakingPool.sol [hunk 3]
```diff
@@ -83,11 +83,11 @@ contract StakingPool {
 
     // Add stakes
     function () external payable {
-        require(minStakes <= msg.value, "Invalid stakes.");
         calculatePayout();
         StakerInfo storage info = stakerInfo[msg.sender];
         // New staker
         if (info.stakes == 0) {
+            require(info.stakes.add(msg.value) >= minStakes, "Invalid stakes.");
```
Comment (qizhou) [inline contracts/StakingPool.sol:90]: If info.stakes == 0, should the code become "require(msg.value >= minStakes, ...)"?

Comment (jishankai) [reply contracts/StakingPool.sol:90]: Yes..

**Pull request merged**